// json-to-java/bin/notations/class/Attribute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const J2JError_1 = tslib_1.__importDefault(require("../../utils/J2JError"));
const json_1 = require("../../utils/json");
const QuickConsole_1 = tslib_1.__importDefault(require("../../utils/QuickConsole"));
const BaseWithName_1 = tslib_1.__importDefault(require("../BaseWithName"));
const Modifier_1 = require("../basic/Modifier");
function parseAttributes(emitter, fieldName, receiver, attributeJson, currentIndent) {
    attributeJson.forEach((attribute, index) => {
        if (json_1.JsonUtil.isJsonObject(attribute)) {
            receiver.push(new JavaClassAttribute(currentIndent, attribute));
        }
        else {
            throw J2JError_1.default.elementTypeError(emitter, fieldName, index, attributeJson.length, Object);
        }
    });
}
exports.parseAttributes = parseAttributes;
class JavaClassAttribute extends BaseWithName_1.default {
    constructor(currentIndent, json) {
        super(currentIndent, json.name);
        this.accessModifier = null;
        this.nonAccessModifiers = [];
        this.value = null;
        this.nameWhenAsEmitter = 'Attribute';
        if ('accessModifier' in json) {
            if (Modifier_1.isJavaAccessModifier(json.accessModifier)) {
                this.accessModifier = json.accessModifier;
            }
            else {
                throw J2JError_1.default.valueNotAccepted(this, 'accessModifier', json.accessModifier);
            }
        }
        if ('nonAccessModifiers' in json) {
            if (json_1.JsonUtil.isJsonArray(json.nonAccessModifiers)) {
                Modifier_1.parseNonAccessModifiers(this, 'nonAccessModifiers', this.nonAccessModifiers, json.nonAccessModifiers);
            }
            else {
                QuickConsole_1.default.warnIgnoreField(this, 'nonAccessModifiers', Array);
            }
        }
        if ('type' in json) {
            if (typeof json.type === 'string') {
                this.type = json.type;
            }
            else {
                throw J2JError_1.default.typeError(this, 'type', String);
            }
        }
        else {
            throw J2JError_1.default.fieldNotDefined(this, 'type');
        }
        if ('value' in json) {
            if (json.value === null || typeof json.value === 'string') {
                this.value = json.value;
            }
            else {
                QuickConsole_1.default.warnIgnoreField('Attributes', 'value', [null, String]);
            }
        }
    }
    toString() {
        return '' +
            `\n${this.currentIndentString}` +
            (this.accessModifier ? `${this.accessModifier} ` : '') +
            this.nonAccessModifiers.join(' ') + (this.nonAccessModifiers.length > 0 ? ' ' : '') +
            this.type + ' ' +
            this.name +
            (this.value !== null ? ` = ${this.value}` : '') +
            ';';
    }
}
exports.default = JavaClassAttribute;
