// json-to-java/bin/notations/class/Constructor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const J2JError_1 = tslib_1.__importDefault(require("../../utils/J2JError"));
const json_1 = require("../../utils/json");
const QuickConsole_1 = tslib_1.__importDefault(require("../../utils/QuickConsole"));
const BaseWithName_1 = tslib_1.__importDefault(require("../BaseWithName"));
const Modifier_1 = require("../basic/Modifier");
const Statement_1 = require("../basic/Statement");
const VariableDifinition_1 = require("../basic/VariableDifinition");
function parseConstructors(emitter, fieldName, className, receiver, constructorJson, currentIndent) {
    constructorJson.forEach((attribute, index) => {
        if (json_1.JsonUtil.isJsonObject(attribute)) {
            receiver.push(new JavaClassConstructor(currentIndent, className, attribute));
        }
        else {
            throw J2JError_1.default.elementTypeError(emitter, fieldName, index, constructorJson.length, Object);
        }
    });
}
exports.parseConstructors = parseConstructors;
class JavaClassConstructor extends BaseWithName_1.default {
    constructor(currentIndent, className, json) {
        super(currentIndent, className);
        this.accessModifier = null;
        this.arguments = [];
        this.statements = [];
        this.nameWhenAsEmitter = 'Constructor';
        if ('accessModifier' in json) {
            if (Modifier_1.isJavaAccessModifier(json.accessModifier)) {
                this.accessModifier = json.accessModifier;
            }
            else {
                throw J2JError_1.default.valueNotAccepted(this, 'accessModifier', json.accessModifier);
            }
        }
        if ('arguments' in json) {
            if (json_1.JsonUtil.isJsonArray(json.arguments)) {
                VariableDifinition_1.parseVariableDifinitions(this, 'arguments', this.arguments, json.arguments, currentIndent);
            }
        }
        if ('statements' in json) {
            if (json_1.JsonUtil.isJsonArray(json.statements)) {
                Statement_1.parseJavaStatements(this.statements, json.statements, currentIndent);
            }
            else {
                QuickConsole_1.default.warnIgnoreField(this, 'statements', Array);
            }
        }
    }
    toString() {
        return '' +
            `\n${this.currentIndentString}` +
            (this.accessModifier ? `${this.accessModifier} ` : '') +
            `${this.name} ` +
            `(${this.arguments.join(', ')}) {` +
            Statement_1.JavaStatementToString(this.statements, depth => this.contentIndentString(depth)) +
            `\n${this.currentIndentString}}`;
    }
}
exports.default = JavaClassConstructor;
